### PR TITLE
Add buffered line reads to BufferedSource

### PR DESCRIPTION
**Pull request by Zaczero**
Provide BufferedSource::readLine for opt-in buffered line reading. Migrate applicable call sites.

Succeeds https://github.com/NixOS/nix/pull/14829

## Context

---

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

**Comment by xokdvium** [comment]
Any chance we could use Pipe class in tests?

**Comment by xokdvium** [comment]
Also could we test this code with different buffer sizes? To catch various corner-cases. Maybe make a parametric test fixture and feed random data (mt19937 maybe?) into the source and test all against an unbuffered version via a StringSource (or does that not take a Source &? In that case we can just di a temp file). This buffer handling is pretty hard, so I want us to get it 100% right

**Comment by Zaczero** [comment]
I added buffer oriented tests

File: src/libutil-tests/file-descriptor.cc [hunk 1]
```diff
@@ -0,0 +1,249 @@
+#include <gtest/gtest.h>
+
+#include "nix/util/file-descriptor.hh"
+#include "nix/util/serialise.hh"
+
+#include <cstring>
+
+#ifndef _WIN32
```
Comment (xokdvium) [inline src/libutil-tests/file-descriptor.cc]: Generally the ifdef should be redundant, since there's a windows-specific implemntation of Pipe (not sure how well tested it is).

File: src/libutil/serialise.cc [hunk 2]
```diff
@@ -139,6 +139,50 @@ size_t BufferedSource::read(char * data, size_t len)
     return n;
 }
 
+std::string BufferedSource::readLine(bool eofOk)
+{
+    if (!buffer)
+        buffer = decltype(buffer)(new char[bufSize]);
```
Comment (xokdvium) [inline src/libutil/serialise.cc]: We could probably use this, but that's just stylistic nitpick and pre-existing, so just a nice-to-have.

```suggestion
        buffer = std::make_unique_for_overwrite<char[]>(bufSize);
```

File: src/libutil/serialise.cc [hunk 3]
```diff
@@ -139,6 +139,50 @@ size_t BufferedSource::read(char * data, size_t len)
...
+
+    std::string line;
+    while (true) {
+        if (bufPosOut < bufPosIn) {
+            auto * start = buffer.get() + bufPosOut;
+            auto * end = buffer.get() + bufPosIn;
+            if (auto * newline = static_cast<char *>(memchr(start, '\n', end - start))) {
```
Comment (xokdvium) [inline src/libutil/serialise.cc:152]: Hope we could wrap this in a helper method that returns a `std::span`. This manual handling of bounds is quite tricky.

File: src/libutil/serialise.cc [hunk 4]
```diff
@@ -139,6 +139,50 @@ size_t BufferedSource::read(char * data, size_t len)
...
+        }
+
+        size_t n = 0;
+        try {
+            n = readUnbuffered(buffer.get(), bufSize);
+        } catch (EndOfFile & e) {
+            bufPosOut = bufPosIn = 0;
+            if (eofOk)
+                return line;
+            throw EndOfFile("unexpected EOF reading a line");
```
Comment (xokdvium) [inline src/libutil/serialise.cc:168]: This and below could maybe get deduplicated with a lambda, but that's short enough to maybe not warrant that.

**Review by xokdvium — APPROVED** [review]

**Comment by xokdvium** [comment]
Fixed those minor nitpicks that I had.
